### PR TITLE
PTC Improvements and BugFixes

### PR DESCRIPTION
**Pull request by jgray-19**
Please see [presentation](https://indico.cern.ch/event/1310491/?view=standard_numbered_inline_minutes) for the list of changes to the code.

Changes to note: 
- A flag has been added to `ptc_setswitch` called `nocharge`.  If you set this value to `false`, PTC receives the true value of the charge from your `beam` rather than 1. This has a significant effect on the tests, where it used to tell the user that the charge that had been "set" was the charge from the `beam`, not the actual charge that was set. Furthermore, in debug mode, if `nocharge` is set is output, therefore the tests have been changed to accommodate this.   
- `ptcrbend` and `truerbend` has been removed. Now we use `true_rbend`, therefore the diamond sequence has been changed accordingly.
- Setting `n_wedge` to 0 is more stable (due to no 0 rotations), however this means that the test fails at on the final digit of the test-ptc-twiss-56Dt-ini_map_man and test-ptc-align tests. These tests have therefore been changed.
- Solenoids are never converted to markers now, as they can have knl and ksl values, therefore they are changed to multipoles when `l=0 and (lrad=0 or ksi=0)`. The tests have now been changed accordingly.
- The inconsistency in sector_nmul_max, reported by Laurent [here](https://indico.cern.ch/event/1286803/), has been corrected. The tests have been changed accordingly
- The solenoid problem described in the [presentation](https://indico.cern.ch/event/1310491/?view=standard_numbered_inline_minutes) has also been fixed (`trsol` of trrun.f90). This is the only changed to the physics of MAD-X, everything else is changes to MAD-X-PTC, including documentation.

**--EDIT--**
I have now removed the changes to the solenoid equation. After running the tests, I found that both versions converged to the thick lens in the solenoid, but the previous implementation converged fastest in MAD-X, this is _not_ the case in PTC or MAD-NG.

Closes issues:
#1160, #1135 and #1165 (was never completed)

### CODE REVIEW

**Comment by rdemaria** [comment]
Are the additional ptc attributes needed for all the elements in madx_dict.c? They have a memory impact.

File: src/trrun.f90 [hunk 1]
```diff
@@ -3615,7 +3615,9 @@ subroutine trsol(track,ktrack,dxt,dyt)
               track(2,i) =  pxf_
               track(3,i) = -xf  * sinTh  +  yf  * cosTh
               track(4,i) =  pyf_
-              track(5,i) =  (sigf + (xf*pyf - yf*pxf)*Z) * beti
+              ! JG: 25.07.2023 - Fix for T variable to match PTC, 
+              ! see http://cds.cern.ch/record/290467/files/arXiv:acc-phys_9510005.pdf
+              track(5,i) =  (sigf + (xf*track(4,i) - yf*track(2,i))*Z) * beti
```
[the remark this answers is not in the document]
Comment (jgray-19) [reply src/trrun.f90:3620]: The variables `pxf_` and `pyf_` are currently being used by this function, could potentially have new variables called `pxi` and `pyi` :

Line from 3550
```
           !     Ripken formulae p.28 (3.35 and 3.36)
           xf    = track(1,i)
           pxi  = track(2,i)
           yf    = track(3,i)
           pyi  = track(4,i)
           psigf = track(6,i) * beti
```
Then it becomes
```suggestion
              track(5,i) =  (sigf + (xf*pyi - yf*pxi)*Z) * beti
```

**Comment by jgray-19** [comment]
> Are the additional ptc attributes needed for all the elements in madx_dict.c? They have a memory impact.

For the elements, quadrupole, sextupole, octupole, rbend and sbend, they all go through the fringe maps, so for activation of fringes, yes. However, if you are concerned about memory, there are some elements that do not have fringe maps, such as the electrostatic separator, in which you could remove current arguments, such as `kill_e(nt|xi)_fringe` or `fringe` (of course, unless this is needed for MAD-X).

**Comment by ldeniau** [comment]
About memory impact, the fringe field attributes use about 8 numbers (let's round it to 10). So for example for the LHC, loading LHCB1 and LHCB2 in memory will use 500 MB (1GB after a matching), while the attribute will use 8 bytes*10 attributes*10000 elements ~ 0.5 MB. If all elements get those attributes (not the case) rounding up all numbers leads to a 0,1% memory increase.

**Comment by rdemaria** [comment]
I see you are still adding important stuff. How much time do you need to complete it?

**Comment by jgray-19** [comment]
Please see pull request #1203. With this fix, we can rerun the failing check, allowing it to pass. If you would like to see that the master fails, you can rerun this test, as I did for https://github.com/jgray-19/MAD-X/tree/master.

**Comment by rdemaria** [comment]
ok I merge!

**Comment by rdemaria** [comment]
can you add your PRs to https://github.com/MethodicalAcceleratorDesign/MAD-X/blob/master/Changes.md ?

**Comment by jgray-19** [comment]
Yes, of course! Will look at this tomorrow.

**Pull request merged**